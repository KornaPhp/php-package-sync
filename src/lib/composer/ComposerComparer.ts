import { ComparisonKind, FileComparisonResult } from '../../types/FileComparisonResult';
import { Composer } from './Composer';
const semver = require('semver');

export class ComposerComparer {
    static compareScripts(skeletonPath: string, repositoryPath: string): FileComparisonResult[] {
        const skeletonComposer = Composer.create(`${skeletonPath}/composer.json`);
        const repositoryComposer = Composer.create(`${repositoryPath}/composer.json`);

        return skeletonComposer
            .scripts()
            .filter(script => !repositoryComposer.hasScript(script.name))
            .map(script => {
                return {
                    kind: ComparisonKind.PACKAGE_SCRIPT_NOT_FOUND,
                    score: 0,
                    name: script.name,
                    context: Object.assign({}, script),
                };
            });
    }

    static comparePackages(skeletonPath: string, repositoryPath: string): FileComparisonResult[] {
        const skeletonComposer = Composer.create(`${skeletonPath}/composer.json`);
        const repositoryComposer = Composer.create(`${repositoryPath}/composer.json`);

        return skeletonComposer
            .packages('all')
            .map(pkg => {
                if (!repositoryComposer.hasPackage(pkg.name, pkg.section)) {
                    return <FileComparisonResult>{
                        kind: ComparisonKind.PACKAGE_NOT_USED,
                        score: 0,
                        name: pkg.name,
                        context: pkg.section,
                    };
                }

                const repositoryPackage = repositoryComposer.package(pkg.name);
                const version1 = semver.coerce(pkg.version);
                const version2 = semver.coerce(repositoryPackage.version);

                if (semver.gt(version1, version2)) {
                    const versionDiff = semver.diff(version1, version2);

                    return {
                        kind: ComparisonKind.PACKAGE_VERSION_MISMATCH,
                        score: versionDiff,
                        name: pkg.name,
                        context: pkg,
                    };
                }

                return <FileComparisonResult>{ name: '!remove' };
            })
            .filter(item => item.name !== '!remove');
    }
}
