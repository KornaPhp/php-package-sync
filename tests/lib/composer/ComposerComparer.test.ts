/* eslint-disable no-undef */

import { ComposerComparer } from '../../../src/lib/composer/ComposerComparer';

it('compares packages between two directories', () => {
    const results = ComposerComparer.comparePackages(
        `${__dirname}/../../data/test-skeleton`,
        `${__dirname}/../../data/test-package`,
    )
        .map(result => ({ kind: result.kind, name: result.name, score: result.score, ctx: result.context }));

    expect(results)
        .toMatchSnapshot();
});

it('compares scripts between two directories', () => {
    const results = ComposerComparer.compareScripts(
        `${__dirname}/../../data/test-skeleton`,
        `${__dirname}/../../data/test-package`,
    )
        .map(result => ({ kind: result.kind, name: result.name, score: result.score, ctx: result.context }));

    expect(results)
        .toMatchSnapshot();
});
