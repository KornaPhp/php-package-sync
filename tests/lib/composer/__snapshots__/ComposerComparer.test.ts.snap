// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compares packages between two directories 1`] = `
Array [
  Object {
    "ctx": Object {
      "name": "psr/http-message",
      "section": "require",
      "version": "^1.5",
    },
    "kind": "pkg_version",
    "name": "psr/http-message",
    "score": "minor",
  },
  Object {
    "ctx": "require",
    "kind": "missing_pkg",
    "name": "spatie/macroable",
    "score": 0,
  },
  Object {
    "ctx": "require-dev",
    "kind": "missing_pkg",
    "name": "vimeo/psalm",
    "score": 0,
  },
]
`;

exports[`compares scripts between two directories 1`] = `
Array [
  Object {
    "ctx": Object {
      "command": "vendor/bin/phpunit --coverage-html coverage",
      "name": "test-coverage",
    },
    "kind": "pkg_script",
    "name": "test-coverage",
    "score": 0,
  },
  Object {
    "ctx": Object {
      "command": "vendor/bin/php-cs-fixer fix --allow-risky=yes",
      "name": "format",
    },
    "kind": "pkg_script",
    "name": "format",
    "score": 0,
  },
]
`;
